physics/cu_unified_deep.F90

```diff
@@ -0,0 +1,6051 @@
+       pret
+!$acc declare copy(outq,outt,outqc,outu,outv,q,pret)
+      character *(*), intent (in)         ::                                  &
+       name
+     real(kind=kind_phys)                                                                     &
+        ,intent (in  )                   ::                                   &
+        dt
+     real(kind=kind_phys) :: names,scalef,thresh,qmem,qmemf,qmem2,qtest,qmem1
+     integer :: icheck
+!
+! first do check on vertical heating rate
+!
+      thresh=300.01
+!      thresh=200.01        !ss
+!      thresh=250.01
+      names=1.
+      if(name == 'shallow' .or. name == 'mid')then
+        thresh=148.01
+        names=1.
+      endif
+      scalef=86400.
+!$acc kernels
+!$acc loop private(qmemf,qmem,icheck)
+      do i=its,itf
+      if(ktop(i) <= 2)cycle
+      icheck=0
+      qmemf=1.
+      qmem=0.
+!$acc loop reduction(min:qmemf)
+      do k=kts,ktop(i)
+         qmem=(outt(i,k))*86400.
+         if(qmem.gt.thresh)then
+           qmem2=thresh/qmem
+           qmemf=min(qmemf,qmem2)
+      icheck=1
+!
+!
+!          print *,'1',' adjusted massflux by factor ',i,j,k,qmem,qmem2,qmemf,dt
+         endif
+         if(qmem.lt.-.5*thresh*names)then
+           qmem2=-.5*names*thresh/qmem
+           qmemf=min(qmemf,qmem2)
+      icheck=2
+!
+!
+         endif
+      enddo
+      do k=kts,ktop(i)
+         outq(i,k)=outq(i,k)*qmemf
+         outt(i,k)=outt(i,k)*qmemf
+         outu(i,k)=outu(i,k)*qmemf
+         outv(i,k)=outv(i,k)*qmemf
+         outqc(i,k)=outqc(i,k)*qmemf
+      enddo
+      pret(i)=pret(i)*qmemf 
+      enddo
+!$acc end kernels
+!      return
+!
+! check whether routine produces negative q's. this can happen, since 
+! tendencies are calculated based on forced q's. this should have no
+! influence on conservation properties, it scales linear through all
+! tendencies
+!
+!      return
+!      write(14,*)'return'
+      thresh=1.e-32
+!$acc kernels
+!$acc loop private(qmemf,qmem,icheck)
+      do i=its,itf
+      if(ktop(i) <= 2)cycle
+      qmemf=1.
+!$acc loop reduction(min:qmemf)
+      do k=kts,ktop(i)
+         qmem=outq(i,k)
+         if(abs(qmem).gt.0. .and. q(i,k).gt.1.e-6)then
+         qtest=q(i,k)+(outq(i,k))*dt
+         if(qtest.lt.thresh)then
+!
+! qmem2 would be the maximum allowable tendency
+!
+           qmem1=abs(outq(i,k))
+           qmem2=abs((thresh-q(i,k))/dt)
+           qmemf=min(qmemf,qmem2/qmem1)
+           qmemf=max(0.,qmemf)
+         endif
+         endif
+      enddo
+      do k=kts,ktop(i)
+         outq(i,k)=outq(i,k)*qmemf
+         outt(i,k)=outt(i,k)*qmemf
+         outu(i,k)=outu(i,k)*qmemf
+         outv(i,k)=outv(i,k)*qmemf
+         outqc(i,k)=outqc(i,k)*qmemf
+      enddo
+      pret(i)=pret(i)*qmemf 
+      enddo
+!$acc end kernels
+   end subroutine neg_check
+
+!> This subroutine calculates final output fields including
+!! physical tendencies, precipitation, and mass-flux.
```

Comment:
@dustinswales @grantfirl @Qingfu-Liu the addition of prognostic closure in the below subroutine is new